pfcpiface/bess.go

```diff
@@ -677,72 +690,151 @@ func (b *bess) processQER(ctx context.Context, any *anypb.Any, method upfMsgType
 
 	methods := [...]string{"add", "add", "delete", "clear"}
 
-	_, err := b.client.ModuleCommand(ctx, &pb.CommandRequest{
+	val, err := b.client.ModuleCommand(ctx, &pb.CommandRequest{
 		Name: "qerLookup",
 		Cmd:  methods[method],
 		Arg:  any,
 	})
 	if err != nil {
 		log.Println("qerLookup method failed!:", err)
 	}
+	log.Println("process QER : ", val)
 }
 
 func (b *bess) addQER(ctx context.Context, done chan<- bool, qer qer) {
 	go func() {
-		var (
-			any *anypb.Any
-			err error
-		)
-
-		q := &pb.ExactMatchCommandAddArg{
-			Gate: uint64(0),
-			Fields: []*pb.FieldData{
-				intEnc(uint64(qer.qerID)), /* far_id */
-				intEnc(qer.fseID),         /* fseid */
-			},
-			Values: []*pb.FieldData{
-				intEnc(uint64(qer.qfi)),      /* action */
-				intEnc(uint64(qer.ulStatus)), /* QFI */
-				intEnc(uint64(qer.dlStatus)), /* tunnel_out_type */
-				intEnc(qer.ulMbr),            /* access-ip */
-				intEnc(qer.dlMbr),            /* enb ip */
-				intEnc(qer.ulGbr),            /* enb teid */
-				intEnc(qer.dlGbr),            /* udp gtpu port */
-			},
+		var any *anypb.Any
+		var err error
+		var cir, pir, cbs, ebs, pbs uint64
+		var srcIface uint8
+		if qosVal, ok := b.qciQosMap[qer.qfi]; ok {
+			cbs = uint64(qosVal.cbs)
+			ebs = uint64(qosVal.ebs)
+			pbs = uint64(qosVal.pbs)
+		} else {
+			log.Println("No config for qfi/qci : ", qer.qfi,
+				". Using default burst size.")
+			cbs = uint64(DefaultBurstSize)
+			ebs = uint64(DefaultBurstSize)
+			pbs = uint64(DefaultBurstSize)
 		}
-
-		any, err = anypb.New(q)
-		if err != nil {
-			log.Println("Error marshalling the rule", q, err)
-			return
+		if qer.ulMbr != 0 {
```

Comment:
Can we remove all these if blocks with

```suggestion
		cir = Min(Max((qer.ulGbr * 1000) / 8, 1), MaxCir)
		pir = Min(Max((qer.ulMbr * 1000) / 8, cir), MaxPir)

```

Reply:
same for DL

Reply:
can remove MaxCIR/MaxPIR. Since mbr/gbr and cir/pir are all 64 bit

```
		cir = Max((qer.ulGbr * 1000) / 8, 1)
		pir = Max((qer.ulMbr * 1000) / 8, cir)
```